pkg/scalers/datadog_scaler.go

```diff
@@ -0,0 +1,316 @@
+package scalers
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/api/autoscaling/v2beta2"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	datadog "github.com/DataDog/datadog-api-client-go/api/v1/datadog"
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+)
+
+type datadogScaler struct {
+	metadata  *datadogMetadata
+	apiClient *datadog.APIClient
+}
+
+type valueType int
+
+const (
+	average = iota
+	global
+)
+
+type datadogMetadata struct {
+	apiKey      string
+	appKey      string
+	datadogSite string
+	query       string
+	queryValue  int
+	vType       valueType
+	metricName  string
+	age         int
+}
+
+var datadogLog = logf.Log.WithName("datadog_scaler")
+
+// NewDatadogScaler creates a new Datadog scaler
+func NewDatadogScaler(ctx context.Context, config *ScalerConfig) (Scaler, error) {
+	meta, err := parseDatadogMetadata(config)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing Datadog metadata: %s", err)
+	}
+
+	apiClient, err := newDatadogConnection(ctx, meta)
+	if err != nil {
+		return nil, fmt.Errorf("error establishing Datadog connection: %s", err)
+	}
+	return &datadogScaler{
+		metadata:  meta,
+		apiClient: apiClient,
+	}, nil
+}
+
+func parseDatadogMetadata(config *ScalerConfig) (*datadogMetadata, error) {
+	meta := datadogMetadata{}
+
+	if val, ok := config.TriggerMetadata["query"]; ok {
+		meta.query = val
+	} else {
+		return nil, fmt.Errorf("no query given")
+	}
+
+	if val, ok := config.TriggerMetadata["queryValue"]; ok {
+		queryValue, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, fmt.Errorf("queryValue parsing error %s", err.Error())
+		}
+		meta.queryValue = queryValue
+	} else {
+		return nil, fmt.Errorf("no queryValue given")
+	}
+
+	if val, ok := config.TriggerMetadata["age"]; ok {
+		age, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, fmt.Errorf("age parsing error %s", err.Error())
+		}
+		meta.age = age
+	} else {
+		meta.age = 90 // Default window 90 seconds
+	}
+
+	// For all the points in a given window, we take the rollup to the window size
+	rollup := fmt.Sprintf(".rollup(avg, %d)", meta.age)
+	meta.query = meta.query + rollup
+
+	if val, ok := config.TriggerMetadata["type"]; ok {
+		val = strings.ToLower(val)
+		switch val {
+		case "average":
+			meta.vType = average
+		case "global":
+			meta.vType = global
+		default:
+			return nil, fmt.Errorf("type has to be global or average")
+		}
+	} else {
+		meta.vType = average // Default to average between pods
+	}
+
+	if val, ok := config.TriggerMetadata["metricName"]; ok {
+		meta.metricName = GenerateMetricNameWithIndex(config.ScalerIndex, kedautil.NormalizeString(fmt.Sprintf("datadog-%s", val)))
+	} else {
+		return nil, fmt.Errorf("no metric name given")
+	}
```

Comment:
Does requiring `metricName` make sense? I mean, `metricName` it's only used to generate the metric name in the scaler, maybe it could be optional and in case of not be provided, generate it in base of the query for example.
There are several examples of this last behavior in other scalers that you could use.

Reply:
+1 

I think that we don't need to enforce the metricName (as it was just for KEDA internal usage). The current implementation ensures that the generated `metricName` is unique, so I think we can drop this one and make the spec simpler.  Unless there's a serious need for it.

Reply:
ah, cool, let me try removing it. And no, there is no need